### ISSUE TITLE
CMake build fails on \gc\base\standard\ParallelGlobalGC.cpp compilation

### ISSUE THREAD

**Issue by samolisov** [eclipse/omr#2404]
Once the commit `3da69438baf4c4cfb6acc89a38dcce716fff6c19` has been merged into the master, `OMR` could not be built using CMake (I use MSVC on Windows), the bellow error appears:

```
..\gc\base\standard\ParallelGlobalGC.cpp
..\gc\base\standard\ParallelGlobalGC.cpp(349): error C2039: '_masterThreadTenureTLHRemainderTop': is not a member of 'MM_GCExtensionsBase'
C:\Neoflex\dev\openj9\omr\gc\base\WorkPackets.hpp(43): note: see declaration of 'MM_GCExtensionsBase'
..\gc\base\standard\ParallelGlobalGC.cpp(350): error C2039: '_masterThreadTenureTLHRemainderBase': is not a member of 'MM_GCExtensionsBase'
C:\Neoflex\dev\openj9\omr\gc\base\WorkPackets.hpp(43): note: see declaration of 'MM_GCExtensionsBase'
```

As I see the field `_masterThreadTenureTLHRemainderTop` is defined in the `GCExtensionBase` class only if the `OMR_GC_MODRON_SCAVENGER` flag is defined:

```cpp
#if defined(OMR_GC_MODRON_SCAVENGER)
	MM_Scavenger *scavenger;
	void *_masterThreadTenureTLHRemainderBase;  /**< base and top pointers of the last unused tenure TLH copy cache, that will be loaded to thread env during master setup */
	void *_masterThreadTenureTLHRemainderTop;
#endif /* OMR_GC_MODRON_SCAVENGER */
```
But the field is used regardless of the flag value:

```cpp
void
MM_ParallelGlobalGC::cleanupAfterGC(MM_EnvironmentBase *env, MM_AllocateDescription *allocDescription)
{
...
	_extensions->_masterThreadTenureTLHRemainderTop = NULL;
	_extensions->_masterThreadTenureTLHRemainderBase = NULL;
}
```

Another one, the class `gc\base\standard\Scavenger.cpp`  contains an invocation of the `oldToOldReferenceCreated` function, the function is defined in the `gc\base\standard\ConcurrentGC` class but the header, as well as the body of the class, will be compiled only if both `OMR_GC_MODRON_CONCURRENT_MARK` and `OMR_GC_MODRON_SCAVENGER` flags are enabled:

```cpp
#if defined(OMR_GC_MODRON_CONCURRENT_MARK)

#include "ConcurrentCardTable.hpp"
...
#if defined(OMR_GC_MODRON_SCAVENGER)
	void scanRememberedSet(MM_EnvironmentBase *env);
	void oldToOldReferenceCreated(MM_EnvironmentBase *env, omrobjectptr_t objectPtr);
#endif /* OMR_GC_MODRON_SCAVENGER */
...
```

so, the linker is unable to find the implementation of the `oldToOldReferenceCreated` function:

```
fvtest\gctest\CMakeFiles\omrgctest.dir/manifest.res" failed (exit code 1120) with the following output:
omrgc.lib(Scavenger.cpp.obj) : error LNK2019: unresolved external symbol oldToOldReferenceCreated referenced in function "public: void __cdecl MM_Scavenger::pruneRememberedSetList(class MM_EnvironmentStandard *)" (?pruneRememberedSetList@MM_Scavenger@@QEAAXPEAVMM_EnvironmentStandard@@@Z)
omrgctest.exe : fatal error LNK1120: 1 unresolved externals
```

In CMake, the flags `OMR_GC_MODRON_SCAVENGER` and `OMR_GC_MODRON_CONCURRENT_MARK` are disabled by default (see cmake/config.cmake):

```cmake
set(OMR_GC_MODRON_CONCURRENT_MARK OFF CACHE BOOL "TODO: Document")
set(OMR_GC_MODRON_SCAVENGER OFF CACHE BOOL "TODO: Document")
```

**Comment by samolisov** [comment]
@RSalman Could you have a look at? Maybe the strings

```cpp
_extensions->_masterThreadTenureTLHRemainderTop = NULL;
_extensions->_masterThreadTenureTLHRemainderBase = NULL;
```

have to be wrapped by  `#if defined(OMR_GC_MODRON_SCAVENGER)` also?

**Comment by RSalman** [comment]
I'll have it fixed.

**Comment by RSalman** [comment]
Opend PR https://github.com/eclipse/omr/pull/2407 for Fix for `ParallelGlobalGC`.

`scanRememberedSet` seems to be fine,  only `oldToOldReferenceCreated` needs to be wrapped. However, I'm not too familiar with  the that code, @amicic will wrapping all calls to `oldToOldReferenceCreated` with `#if defined(OMR_GC_MODRON_SCAVENGER)` be fine for this? If so, I can make the changes and add them to the PR.

**Comment by amicic** [comment]
Call sites for oldToOldReferenceCreated in Scavenger.cpp should be wrapped with OMR_GC_MODRON_CONCURRENT_MARK. No additional wrapping is needed with OMR_GC_MODRON_SCAVENGER (the whole file is already wrapped). Things look ok on ConcurrentGC side, but compiler has the final word :)

**Comment by amicic** [comment]
I have a feeling that a few more things are/were not quite right (even before Salman's change):

- MM_Scavenger should not be part of GCExtensionsBase
- just a few lines before the compiler's complaint there is this line:
MM_EnvironmentStandard *threadEnvironment = MM_EnvironmentStandard::getEnvironment(walkThread);
what compiler is not complaining about

If SCAVENGER and CONCURRENT_MARK are off, ParallelGlobalGC should not be even including MM_EnvironmentStandard. It does somehow (indirectly), otherwise this should have failed too.

**Comment by charliegracie** [comment]
I believe this has been resolved via #2407
